flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java

```diff
@@ -128,6 +135,53 @@ public FlinkKafkaConsumer010(List<String> topics, KeyedDeserializationSchema<T>
 	}
 
 	@Override
+	public FlinkKafkaConsumerBase<T> setStartFromSpecificDate(Date date) {
+		Preconditions.checkArgument(null != date && date.getTime() <= System.currentTimeMillis(), "Startup time must before curr time.");
+		this.startupMode = StartupMode.SPECIFIC_TIMESTAMP;
+		this.specificStartupDate = date;
+		this.specificStartupOffsets = null;
+		return this;
+	}
+
+	/**
+	 * Convert flink topic partition to kafka topic partition.
+	 * @param flinkTopicPartitionMap
+	 * @return
+	 */
+	private Map<TopicPartition, Long> convertFlinkToKafkaTopicPartition(Map<KafkaTopicPartition, Long> flinkTopicPartitionMap) {
+		Map<TopicPartition, Long> topicPartitionMap = new HashMap<>(flinkTopicPartitionMap.size());
+		for (Map.Entry<KafkaTopicPartition, Long> entry : flinkTopicPartitionMap.entrySet()) {
+			topicPartitionMap.put(new TopicPartition(entry.getKey().getTopic(), entry.getKey().getPartition()), entry.getValue());
+		}
+
+		return topicPartitionMap;
+
+	}
+
+	/**
+	 * Search offset from timestamp for each topic in kafka. If no offset exist, use the latest offset.
+	 * @param partitionTimesMap Kafka topic partition and timestamp
+	 * @return Kafka topic partition and the earliest offset after the timestamp. If no offset exist, use the latest offset in kafka
+	 */
+	private Map<KafkaTopicPartition, Long> convertTimestampToOffset(Map<KafkaTopicPartition, Long> partitionTimesMap) {
```

Comment:
I think we need to move this conversion logic to `KafkaConsumerThread`, otherwise we would be instantiating a KafkaConsumer just for the sake of fetching timestamp-based offsets.
That's where the actual "`KafkaTopicPartitionStateSentinel` to actual offset" conversions take place.
See `KafkaConsumerThread` lines 369 - 390
```
// offsets in the state of new partitions may still be placeholder sentinel values if we are:
//   (1) starting fresh,
//   (2) checkpoint / savepoint state we were restored with had not completely
//       been replaced with actual offset values yet, or
//   (3) the partition was newly discovered after startup;
// replace those with actual offsets, according to what the sentinel value represent.
for (KafkaTopicPartitionState<TopicPartition> newPartitionState : newPartitions) {
	if (newPartitionState.getOffset() == KafkaTopicPartitionStateSentinel.EARLIEST_OFFSET) {
		consumerCallBridge.seekPartitionToBeginning(consumerTmp, newPartitionState.getKafkaPartitionHandle());
		newPartitionState.setOffset(consumerTmp.position(newPartitionState.getKafkaPartitionHandle()) - 1);
	} else if (newPartitionState.getOffset() == KafkaTopicPartitionStateSentinel.LATEST_OFFSET) {
		consumerCallBridge.seekPartitionToEnd(consumerTmp, newPartitionState.getKafkaPartitionHandle());
		newPartitionState.setOffset(consumerTmp.position(newPartitionState.getKafkaPartitionHandle()) - 1);
	} else if (newPartitionState.getOffset() == KafkaTopicPartitionStateSentinel.GROUP_OFFSET) {
		// the KafkaConsumer by default will automatically seek the consumer position
		// to the committed group offset, so we do not need to do it.

		newPartitionState.setOffset(consumerTmp.position(newPartitionState.getKafkaPartitionHandle()) - 1);
	} else {
		consumerTmp.seek(newPartitionState.getKafkaPartitionHandle(), newPartitionState.getOffset() + 1);
	}
}
```